### ISSUE TITLE
ytdl.getInfo - Error: Status code: 404 

### ISSUE THREAD

**Issue by Monabr** [fent/node-ytdl-core#923]
The problem is that if i try get info from some urls i get this error. Example below. 

```
const customVideoInfo = await ytdl.getInfo("http://www.youtube.com/watch?v=3Z61pOtkY7w")

Error: Status code: 404
    at ClientRequest.<anonymous> (/workspace/node_modules/miniget/dist/index.js:211:27)
    at Object.onceWrapper (events.js:421:26)
    at ClientRequest.emit (events.js:314:20)
    at ClientRequest.EventEmitter.emit (domain.js:506:15)
    at HTTPParser.parserOnIncomingClient (_http_client.js:601:27)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:122:17)
    at HTTPParser.execute (<anonymous>)
    at TLSSocket.socketOnData (_http_client.js:474:22)
    at TLSSocket.emit (events.js:314:20)
    at TLSSocket.EventEmitter.emit (domain.js:506:15) 
```

**Comment by WaqasIbrahim** [comment]
I have some across this issue (Mainly on age restricted videos). This is an issue with fallback info endpoint "https://www.youtube.com/get_video_info".

https://www.youtube.com/get_video_info?video_id=3Z61pOtkY7w
This results in 404.

Adding a query parameter "html5=1" solves this issue.
https://www.youtube.com/get_video_info?video_id=3Z61pOtkY7w&html5=1

Needs more testing though.

**Comment by Okervill** [comment]
I'm getting the same error with a large number of music videos. These videos don't seem to be age restricted since they can be viewed without logging in. Most recent examples i've found:
https://www.youtube.com/watch?v=C3a1sz2Bp_E
https://www.youtube.com/watch?v=RGHf1yk0m9I
https://www.youtube.com/watch?v=SC4xMk98Pdc

**Comment by gatecrasher777** [comment]
The get_video_info page is no more it seems.  I was relying on this, so had to do a bit of research today. 

ytdl.getInfo should still work using the watch HTML.  Falling back onto the get_info_page will fail though.  

YT has moved that get_info functionality to their INNERTUBE API which returns just the JSON data (responseContext) without the UrlEncoded wrapper.

Getting it requires a post request to https://www.youtube,com/youtubei/v1/player?key=YOUR_KEY

with videoId as a parameter along with some other innertube parameters (and the key) which are easy enough to pick out with any web request to youtube.  It also requires a 'x-goog-visitor-id' in the header, which again you can pick out from the parameters.  

This now conforms closely to their INNERTUBE API channel browse and search post requests.

I have working code that can do this now, if anyone is interested and can't figure it out.

**Comment by rjdg14** [comment]
I've been having this issue in FreeTube recently on some age restricted videos and a few music videos, which uses this code for retrieving its content. I'm hoping the problem with the algorithm will be fixed in the next few days.

**Comment by gatecrasher777** [comment]
Currently ytdl is dependent on the watch page (getWatchHTMLPage), which works in most cases.  

Fallback to getWatchJSONPage fails because it now requires a POST request, rather than a GET request.

Fallback to getVideoInfoPage fails because it now requires the html5=1 fix.

**Comment by vaaski** [comment]
not sure how to use miniget to make a POST request, but #924 adds the mentioned html5 param to getVideoInfoPage and also changes the ID of the age restricted video test, which passes now.

**Comment by WaqasIbrahim** [comment]
> Fallback to getWatchJSONPage fails because it now requires a POST request, rather than a GET request.

Can you give more details about this?

**Comment by gatecrasher777** [comment]
I wrote a bandwidth analyzer implementing the new requestCallback and I noticed that the getWatchJSONPage always fails, notably it is invoked in the test case (VideoId B3eAMGXFw1o). Each request returns about 25 bytes only (3 attempts) and then the pipeline moves on to the  getVideoInfoPage.  

Checking out this video in the browser, I noticed that there is an age verification command in the INNERTUBE_API, following which,  the getWatchJSONPage is invoked with a POST request.  In the above case:

https://www.youtube.com/watch?v=B3eAMGXFw1o&has_verified=1&pbj=1

requires postData as  follows:

```
const params = {
    'command': {
      clickTrackingParams: INNERTUBE_CONTEXT.clickTracking.clickTrackingParams,
      commandMetadata: {
        webCommandMetadata: {
          url: "/watch?v="+id+"&has_verified=1",
          webPageType: "WEB_PAGE_TYPE_UNKNOWN",
          rootVe: 83769
        }
      },
      urlEndpoint: {
        url: "/watch?v="+id+"&has_verified=1"
      }
    },
    'session_token': XSRF_TOKEN
  }
  const postData = querystring.stringify(params);
```
and 

    'content-type':'application/x-www-form-urlencoded',
    'content-length': Buffer.byteLength(postData,'utf8'),

in the headers.  I have got this to work, such that the media formats are returned in the response. 

I'm not sure what rootVe is, or if it ever changes or needs to change.   INNERTUBE_CONTEXT.clickTracking.clickTrackingParams and XSRF_TOKEN can be scraped from the ytcfg.set() in the initial getWatchHTMLPage.

I can perhaps add the bandwidth analyzer to examples if there is interest. It runs the four test cases when run without an argument, or it will analyze the getInfo pipeline for any specific video.  It shows each request url with uncompressed/compressed bandwidth. And whether the getInfo process was successful.

**Comment by gatecrasher777** [comment]
[github.com/gatecrasher777/node-ytdl-core](https://github.com/gatecrasher777/node-ytdl-core)

I put getInfoBandwidth.js in example - you may need to change the require depending on where you run it. 

getWatchJSONPage, and getVideoInfoPage  are working in the lib/info.js

Also added a fourth info source into the pipeline from the innertube api, called getVideoInfoInner.  Probably superfluous, but it works well and uses very little bandwidth, once you have the ytcfg params/tokens.

**Comment by rjdg14** [comment]
Does anyone know when the module will be updated to fix this issue? It looks like there are a number of working fixes (from the sound of it) but to the best of my knowledge none of them have been properly implemented into an official release yet.

**Comment by vaaski** [comment]
@rjdg14 #924 was merged and released in 4.8.1 and it fixes age-restricted videos.

**Comment by gatecrasher777** [comment]
The getWatchJSONPage still doesn't work.  Ordinarily this would resolve the age restricted video - before getInfo() falls back onto getVideoInfoPage. Perhaps getVideoInfoPage with html5=1 doesn't resolve the age restriction problem in every case. 

Also with getWatchJSONPage being broken,  there are several fruitless attempts to get it which significantly slows down the pipeline.  To address this delay one could change the pipeline order on line 61 to 

```
let info = await pipeline([id, options], validate, retryOptions, [
    getWatchHTMLPage,
    getVideoInfoPage, 
    getWatchJSONPage
]);
```

I have written a fix (mentioned above) but it is quite a major change to the current code, so I'm not willing to make a PR of it before a code maintainer advises me to, or has made suggestions/changes.

**Comment by xpiREC** [comment]
It seems that Google has updated the verification procedure, now the scent cookie file for logged-in users is also taken into account.

**Comment by rjdg14** [comment]
The issue with playback on age restricted videos through the ytdl core seems to have started occuring again, only a couple of weeks after the fix. Should it be possible to fix the issue again?

**Comment by gatecrasher777** [comment]
The getVideoInfoPage (with html=1) would usually return streams even to unverified users.  Maybe this has changed again. 

If you  do supply the cookie of a logged in user to getInfo then  getWatchHTMLPage and getWatchJSONPage should get you past the barrier and receiving the streams without the need to call getVideoInfoPage.  

Except that currently getWatchJSONPage doesn't work. It now requires a post request rather than a get request.

Perhaps getting this working again will help.

**Comment by rjdg14** [comment]
I have an age verified Google account, but still dislike receiving the viewer discretion warning screens on such videos when viewed on YouTube's website, so prefer to use FreeTube which doesn't display them.

**Comment by WaqasIbrahim** [comment]
Related issues:

https://github.com/ytdl-org/youtube-dl/issues/29333
https://github.com/ytdl-org/youtube-dl/issues/29086

There is another solution proposed here but not sure how long this will last as YouTube is moving to new innertube API.
https://github.com/ytdl-org/youtube-dl/issues/29333#issuecomment-864049544

**Comment by JeHwanYoo** [comment]
I think I found the problem.

<img src="https://i.ibb.co/CbW9BkN/2021-06-22-9-14-01.png" alt="2021-06-22-9-14-01" border="0">

I tracked functions in the library, and when I took a log, the URL was strange...

plz check `function getVideoInfoPage()` in ytdl-core/lib/info.js

**Comment by TimeForANinja** [comment]
🤔 might have been some early signs of https://github.com/fent/node-ytdl-core/issues/939

**Comment by TimeForANinja** [comment]
gonna close this as fixed with #924 and #939
feel free to comment again if i'm wrong
actually, might as well just open a new issue in that case😅
